aviatrix/resource_aviatrix_app_domain.go

```diff
@@ -0,0 +1,237 @@
+			tags := make(map[string]string)
+			for key, value := range selectorInfo["tags"].(map[string]interface{}) {
+				tags[key] = value.(string)
+			}
+			filter.Tags = tags
+		}
+
+		appDomain.Selector.Expressions = append(appDomain.Selector.Expressions, filter)
+	}
+
+	return appDomain, nil
+}
+
+func resourceAviatrixAppDomainCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*goaviatrix.Client)
+
+	appDomain, err := marshalAppDomainInput(d)
+	if err != nil {
+		return diag.Errorf("failed to marshal inputs for App Domain during create: %s", err)
+	}
+
+	flag := false
+	defer resourceAviatrixAppDomainReadIfRequired(ctx, d, meta, &flag)
+
+	uuid, err := client.CreateAppDomain(ctx, appDomain)
+	if err != nil {
+		return diag.Errorf("failed to create App Domain: %s", err)
+	}
+	d.SetId(uuid)
+	return resourceAviatrixAppDomainReadIfRequired(ctx, d, meta, &flag)
+}
+
+func resourceAviatrixAppDomainReadIfRequired(ctx context.Context, d *schema.ResourceData, meta interface{}, flag *bool) diag.Diagnostics {
+	if !(*flag) {
+		*flag = true
+		return resourceAviatrixAppDomainRead(ctx, d, meta)
+	}
+	return nil
+}
+
+func resourceAviatrixAppDomainRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*goaviatrix.Client)
+
+	uuid := d.Id()
+	d.Set("uuid", uuid)
+
+	appDomain, err := client.GetAppDomain(ctx, uuid)
+	if err != nil {
+		if err == goaviatrix.ErrNotFound {
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("failed to read App Domain: %s", err)
+	}
+
+	d.Set("name", appDomain.Name)
+
+	var expressions []interface{}
+
+	for _, filter := range appDomain.Selector.Expressions {
+		filterMap := map[string]interface{}{
+			"type":         filter.Type,
+			"cidr":         filter.CIDR,
+			"res_id":       filter.ResId,
+			"account_id":   filter.AccountId,
+			"account_name": filter.AccountName,
+			"region":       filter.Region,
+			"zone":         filter.Zone,
+			"tags":         filter.Tags,
+		}
+
+		expressions = append(expressions, filterMap)
+	}
+
+	selector := []interface{}{
+		map[string]interface{}{
+			"match_expressions": expressions,
+		},
+	}
+	if err := d.Set("selector", selector); err != nil {
+		return diag.Errorf("failed to set selector during App Domain read: %s", err)
+	}
+
+	return nil
+}
+
+func resourceAviatrixAppDomainUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*goaviatrix.Client)
+
+	uuid := d.Id()
+	d.Partial(true)
+	if d.HasChanges("name", "selector") {
+		appDomain, err := marshalAppDomainInput(d)
+		if err != nil {
+			return diag.Errorf("failed to marshal inputs for App Domain during update: %s", err)
+		}
+
+		err = client.UpdateAppDomain(ctx, appDomain, uuid)
+		if err != nil {
+			return diag.Errorf("failed to update App Domain selector: %s", err)
+		}
+	}
+
+	d.Partial(false)
+	return resourceAviatrixAppDomainRead(ctx, d, meta)
```

Comment:
I didn't make it clear, we should use resourceAviatrixAppDomainReadIfRequired() here. Same for the Update() for the other resource.

Reply:
I don't think we normally do ReadIfRequired for Update. It makes sense but we can definitely talk about this.